aten/src/ATen/SparseTensorImpl.cpp

```diff
@@ -62,10 +62,7 @@ TensorImpl* SparseTensorImpl::maybe_zero_dim(bool condition_when_zero_dim) {
            " changing dimensionality via maybe_zero_dim");
   return this;
 }
-std::unique_ptr<Storage> SparseTensorImpl::storage() {
-  AT_ERROR("sparse tensors do not have storage");
-}
-at::StorageImpl* SparseTensorImpl::storageImpl() const {
+const Storage& SparseTensorImpl::storage() {
```

Comment:
I don't really buy that there's no downside here -- the user needs to understand how storages are ref-counted, otherwise they can hold on to a dangling references.  I guess it's fine at the Impl level where presumably people understand what is going on, but at the Tensor level it's not obvious to me that that is the case.

Reply:
Offline convo:

Storage will never dangle as long as Tensor is live. Instead, you'd get the following behavior:

```
Storage s = x.storage(); // equivalent: auto s = ...
x.set_(y.storage());
// s still points to the original storage of x (before set_)

// versus

const Storage& s = x.storage();
x.set_(y.storage());
// s now points to y's storage
```

This is idiomatic C++ behavior, and the non-reference case (which is default) matches PyTorch behavior in Python:

```
>>> x = torch.zeros(2)
>>> y = torch.ones(2)
>>> s = x.storage()
>>> s
 0.0
 0.0
[torch.FloatStorage of size 2]
>>> x.set_(y.storage())
tensor([1., 1.])
>>> s
 0.0
 0.0
[torch.FloatStorage of size 2]
>>> x
tensor([1., 1.])
```